codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/aws/DeserializeMiddleware.java

```diff
@@ -0,0 +1,219 @@
+                        "handler", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("DeserializeHandler"),
+                        "output", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("DeserializeOutput"),
+                        "metadata", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("Metadata"),
+                        "opName", deserialName,
+                        "body", generateHandleDeserializeBody())
+        );
+    }
+
+    private GoWriter.Writable generateHandleDeserializeBody() {
+        return goTemplate("""
+                $errors:W
+
+                $response:W
+            """,
+                MapUtils.of(
+                        "errors", handleResponseChecks(),
+                        "response", handleResponse()
+                ));
+    }
+
+
+    private GoWriter.Writable handleResponseChecks() {
+        return goTemplate("""
+
+                out, metadata, err = next.HandleDeserialize(ctx, in)
+                if err != nil {
+                    return out, metadata, err
+                }
+
+                resp, ok := out.RawResponse.($response:P)
+                if !ok {
+                    return out, metadata, $errorf:T("unexpected transport type %T", out.RawResponse)
+                }
+
+                if resp.Header.Get("smithy-protocol") != $protocol:S {
+                    return out, metadata, &$deserError:T{
+                        Err: $errorf:T(
+                            "unexpected smithy-protocol response header '%s' (HTTP status: %s)",
+                            resp.Header.Get("smithy-protocol"),
+                            resp.Status,
+                        ),
+                    }
+                }
+
+                if resp.StatusCode != 200 {
+                    return out, metadata, &$deserError:T{}
+                }
+
+            """,
+                MapUtils.of(
+                        "response", SMITHY_HTTP_TRANSPORT.pointableSymbol("Response"),
+                        "errorf", GoStdlibTypes.Fmt.Errorf,
+                        "deserError", SmithyGoDependency.SMITHY.struct("DeserializationError"),
+                        "protocol", SMITHY_PROTOCOL_NAME
+                ));
+    }
+
+    private GoWriter.Writable handleResponse() {
+        if (output.members().isEmpty()) {
+            return discardDeserialize();
+        } else if (hasEventStream(ctx.getModel(), output)) {
+            return deserializeEventStream();
+        }
+        return handlePayload();
+    }
+
+    private GoWriter.Writable handlePayload() {
+        return goTemplate("""
+                payload, err := $readAll:T(resp.Body)
+                if err != nil {
+                    return out, metadata, err
+                }
+
+                if len(payload) == 0 {
+                    out.Result = &$output:T{}
+                    return out, metadata, nil
+                }
+
+                decoder := $decoder:T(resp.Body)
+                var cv map[string]interface{}
```

Comment:
I assume this `cv` came from CBOR - but we're working with JSON here. I'd call this either `jv` or just `v`